keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,752 @@
+- `k8s.io/kubernetes/pkg/apis/apps/v1`: `9 December 2023` - `71.4%` <!--(extension of Deployment and ReplicaSet API)-->
+- `k8s.io/kubernetes/pkg/apis/apps/validation`: `9 December 2023` - `92.3%` <!--(validation of the PodReplacementPolicy)-->
+- `k8s.io/kubernetes/pkg/controller/deployment`: `9 December 2023` - `61.7%` <!--(implementation of the PodReplacementPolicy)-->
+- `k8s.io/kubernetes/pkg/controller/deployment/util`: `9 December 2023` - `50.1%` <!--(implementation of the PodReplacementPolicy)-->
+- `k8s.io/kubernetes/pkg/controller/replicaset`: `9 December 2023` - `78.9%` <!--(implementation of the counting of the TerminatingReplicas)-->
+- `k8s.io/kubernetes/pkg/controller`: `9 December 2023` - `71.2%` <!--(creating or updating the utility functions in controller_utils.go)-->
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+Integration tests covering:
+
+<!-- test/integration/replicaset/replicaset_test.go -->
+ReplicaSet
+- The current behavior remains unchanged when the DeploymentPodReplacementPolicy feature gate is disabled.
+- Add a new test that correctly counts `.status.terminatingReplicas` when the DeploymentPodReplacementPolicy feature gate is enabled.
+
+- <test>: <link to test coverage>
+
+<!-- test/integration/deployment/deployment_test.go -->
+Deployment
+- The current behavior remains unchanged when the DeploymentPodReplacementPolicy feature gate is disabled or PodReplacementPolicy is nil.
+- Add a test wrapper for any relevant tests, to ensure that they are run with all possible PodReplacementPolicy values correctly.
+  The relevant tests are those that expect some behavior on Pod deletion, and are affected by this change.
+- Add new tests that observe rollout and scaling transitions for all possible PodReplacementPolicy values and
+  ensure that `.status.terminatingReplicas` is correctly counted when the DeploymentPodReplacementPolicy feature gate is enabled.
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+<!--
+test/e2e/apps/replica_set.go
+test/e2e/upgrades/apps/replica_sets.go
+
+test/e2e/apps/deployment.go
+test/e2e/upgrades/apps/deployments.go
+test/e2e/storage/utils/deployment.go
+-->
+
+- Update all existing e2e tests to account for `.status.TerminatingReplicas` field in both ReplicaSets and Deployments.
```

Comment:
I think the e2e tests may need to be for beta. So you may want to mention that here.

E2E tests can't change feature toggles AFAIK so you should not have any `TerminatingReplicas` until the feature toggle is on by default.

Reply:
Allright, updated the targeting.

Reply:
@aojea has pointed out to me that I was mistaken on when you should add e2e tests so maybe we should mention that in alpha actually. Sorry about that!

Reply:
@kannon92 ack, I have moved the e2es back to alpha